proxy/fuse/fuse.go

```diff
@@ -93,98 +80,94 @@ func NewConnSrc(mountdir, tmpdir string, client *proxy.Client, connset *proxy.Co
-		// Closing the root will ensure the 'dst' chan is closed correctly to
-		// signify that no new connections are possible.
-		if err := root.Close(); err != nil {
-			logging.Errorf("root.Close() error: %v", err)
+	return conns, fuseCloser(func() error {
+		err := srv.Unmount() // Best effort unmount
+		if err != nil {
+			logging.Errorf("Unmount failed: %v", err)
 		}
-		logging.Infof("FUSE exited")
-	}()
-
-	return conns, root, nil
+		return root.Close()
+	}), nil
 }
 
-// symlink implements a symbolic link, returning the underlying string when
-// Readlink is called.
-type symlink string
+type fuseCloser func() error
 
-var _ interface {
-	fs.Node
-	fs.NodeReadlinker
-} = symlink("")
+func (fc fuseCloser) Close() error {
+	return fc()
+}
 
-func (s symlink) Readlink(context.Context, *fuse.ReadlinkRequest) (string, error) {
-	return string(s), nil
+// symlink implements a symbolic link, returning the underlying path when
+// Readlink is called.
+type symlink struct {
+	fs.Inode
+	path string
 }
 
-// Attr helps implement fs.Node.
-func (symlink) Attr(ctx context.Context, a *fuse.Attr) error {
-	*a = fuse.Attr{
-		Mode: 0777 | os.ModeSymlink,
-	}
-	return nil
+var _ fs.NodeReadlinker = &symlink{}
+
+func (s *symlink) Readlink(ctx context.Context) ([]byte, syscall.Errno) {
+	return []byte(s.path), fs.OK
 }
 
+// fsRoot provides the in-memory file system that supports lazy connections to
+// Cloud SQL instances.
 type fsRoot struct {
-	tmpDir, linkDir string
+	fs.Inode
+
+	// tmpDir defines a temporary directory where all the sockets are placed
+	// faciliating connections to Cloud SQL instances.
+	tmpDir string
+	// linkDir is the directory that holds symbolic links to the tmp dir for
+	// each Cloud SQL instance connection. After shutdown, this directory is
+	// cleaned out.
+	linkDir string
 
 	client  *proxy.Client
 	connset *proxy.ConnSet
 
 	// sockLock protects fields in this struct related to sockets; specifically
 	// 'links' and 'closers'.
 	sockLock sync.Mutex
-	links    map[string]symlink
-	// closers holds a slice of things to close when fsRoot.Close is called.
+	links    map[string]*symlink
+	// closers includes a reference to all open Unix socket listeners. When
+	// fs.Close is called, all of these listeners are also closed.
 	closers []io.Closer
 
 	sync.RWMutex
 	dst chan<- proxy.Conn
 }
 
-// Ensure that fsRoot implements the following interfaces.
 var _ interface {
-	fs.FS
-	fs.Node
-	fs.NodeRequestLookuper
-	fs.HandleReadDirAller
+	fs.InodeEmbedder
+	fs.NodeGetattrer
+	fs.NodeLookuper
+	fs.NodeReaddirer
 } = &fsRoot{}
 
 func (r *fsRoot) newConn(instance string, c net.Conn) {
 	r.RLock()
 	// dst will be nil if Close has been called already.
 	if ch := r.dst; ch != nil {
-		ch <- proxy.Conn{instance, c}
+		ch <- proxy.Conn{Instance: instance, Conn: c}
 	} else {
 		logging.Errorf("Ignored new conn request to %q: system has been closed", instance)
 	}
 	r.RUnlock()
 }
 
-func (r *fsRoot) Forget() {
```

Comment:
Just to confirm - fsRoot isn't exported, so this shouldn't be a breaking change?